packages/compass-connections/src/stores/active-connections.ts

```diff
@@ -0,0 +1,76 @@
+import type { ConnectionInfo } from '@mongodb-js/connection-info';
+import { useCallback, useEffect, useState } from 'react';
+import { BSON } from 'bson';
+import {
+  useConnectionRepositoryContext,
+  useConnectionStorageContext,
+} from '@mongodb-js/connection-storage/provider';
+import {
+  ConnectionsManagerEvents,
+  useConnectionsManagerContext,
+} from '../provider';
+import isEqual from 'lodash/isEqual';
+import { ConnectionStorageEvents } from '@mongodb-js/connection-storage/renderer';
+
+/**
+ * Same as _.isEqual, except it takes key order into account
+ */
+function areConnectionsEqual(
+  listA: ConnectionInfo[],
+  listB: ConnectionInfo[]
+): boolean {
+  return isEqual(
+    listA.map((a: any) => BSON.serialize(a)),
+    listB.map((b: any) => BSON.serialize(b))
+  );
+}
+
+export function useActiveConnections(): ConnectionInfo[] {
+  // TODO(COMPASS-7397): services should not be used directly in render method,
+  // when this code is refactored to use the hadron plugin interface, storage
+  // should be handled through the plugin activation lifecycle
+  const connectionManager = useConnectionsManagerContext();
+  const connectionRepository = useConnectionRepositoryContext();
+  const connectionStorage = useConnectionStorageContext();
+
+  const [activeConnections, setActiveConnections] = useState<ConnectionInfo[]>(
+    []
+  );
+
+  const updateList = useCallback(async () => {
+    const list = [
+      ...(await connectionRepository.listFavoriteConnections()),
+      ...(await connectionRepository.listNonFavoriteConnections()),
+    ].filter(({ id }) => connectionManager.statusOf(id) === 'connected');
+    if (!areConnectionsEqual(activeConnections, list)) {
+      setActiveConnections(list);
+    }
+  }, [activeConnections]);
```

Comment:
When you're dealing with React setState that is derived from the previous state, the guideline is to use the setState callback instead of current value from the render scope, this guarantees you are avoiding any chance of running your checks against the stale value (there are also actual hook deps missing here, even though they are stable)

```suggestion
  const updateList = useCallback(async () => {
    const newList = [
      ...(await connectionRepository.listFavoriteConnections()),
      ...(await connectionRepository.listNonFavoriteConnections()),
    ].filter(({ id }) => connectionManager.statusOf(id) === 'connected');
    setActiveConnections(prevList => {
      return areConnectionsEqual(prevList, newList) ? prevList : newList;
    });
  }, [connectionRepository, connectionManager]);
```

Reply:
thank you!